Add basic initial developer docs for TSDB

Dieterbe opened the pull request:
There's a decent amount of content already out there (blog posts,
conference talks, etc), but:
* when they get stale, they don't tend to get updated
* they still leave me with questions that I'ld like to answer
  for developers (like me) who want to use, or work with, TSDB

What I propose is developer docs inside the prometheus
repository.  Easy to find and harness the power of the community
to expand it and keep it up to date.

* perfect is the enemy of good.  Let's have a base and incrementally improve
* Markdown docs should be broad but not too deep.  Source code comments
  can complement them, and are the ideal place for implementation details.

Signed-off-by: Dieter Plaetinck <dieter@grafana.com>

<!--
    Don't forget!

    - Please sign CNCF's Developer Certificate of Origin and sign-off your commits by adding the -s / --sign-off flag to `git commit`. See https://github.com/apps/dco for more information.

    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.

    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.

    - No tests are needed for internal implementation changes.

    - Performance improvements would need a benchmark test to prove it.

    - All exposed objects should have a comment.

    - All comments should start with a capital letter and end with a full stop.
 -->

beorn7 commented:
I leave the details to more @codesome reviewers… but perhaps it would make sense to explain the problem of Go Modules picking up the semantic versioning tag of the prometheus/prometheus repo and wrongly interpreting it as a versioning of Prometheus as a library? I.e. mention the workaround with pinning to a certain commit, something like https://github.com/prometheus/prometheus/issues/7663#issuecomment-663672223 (I'm actually not sure what Cortex maintainers do these days to work around the issue.)

roidelapluie commented:
Yes this is a major annoyance. One of the big issue for developers is that the godoc website is completely out of date, we should mention it as well.

LeviHarrison commented:
I've found [Go Walker](https://gowalker.org) to be a nice solution for this. It generates the documentation directly from main, not tied to a Go Modules version. For example: https://gowalker.org/github.com/prometheus/prometheus/tsdb.

codesome commented on tsdb/docs/usage.md:
```diff
@@ -0,0 +1,152 @@
+* Use Commit() to update the WAL.
+* create a new appender each time you commit.
+
+You can use multiple appenders concurrently (TODO: use case? trade-offs ?)
```
Yes

codesome commented on tsdb/head.go:
```diff
@@ -1451,17 +1456,17 @@ type memSeries struct {
+	headChunk     *memChunk       // Most recent chunk in memory that's still being built.
 	chunkRange    int64
-	firstChunkID  int
+	firstChunkID  int // ChunkID corresponding to mmappedChunks[0].ref
```
That's not true. It refers to the ChunkID for mmappedChunks[0], and mmappedChunks[1] being firstChunkID+1, and so on. With firstChunkID+len(mmappedChunks) being the head chunk.

Dieterbe replied:
I think what you're saying is a superset of what i'm saying?
What's wrong with my comment? - can you make a suggestion?

codesome replied:
Not really. ChunkID referes to the id that corresponds to `mmappedChunks[0]`, and `firstChunkID+i` for `mmappedChunks[i]`, `mmappedChunks[0].ref` is the chunk reference used to access m-mapped chunk from the disk (the 4/4B packed ref).

codesome commented on tsdb/head.go:
```diff
@@ -1451,17 +1456,17 @@ type memSeries struct {
-	sampleBuf     [4]sample
-	pendingCommit bool // Whether there are samples waiting to be committed to this series.
+	nextAt        int64     // Timestamp at which to cut the next chunk.
+	sampleBuf     [4]sample // We keep the last 4 samples here (in addition to appending them) so we don't need coordination between appender and querier. Probably because even the most compact encoding of a sample takes 2 bits, so the last byte is not contended.
```
Right about the sentence starting `Probably...` (but would not keep it in the code)
```suggestion
	// We keep the last 4 samples here (in addition to appending them to the chunk) so we don't
	// need coordination between appender and querier for the chunk bytes that are still being written to.
	sampleBuf     [4]sample
```

Dieterbe replied:
> Right about the sentence starting Probably... (but would not keep it in the code)

I cleaned it up, but i think it's super useful to keep this in the code (without the "probably"), the question of "why do we keep exactly 4 samples in a separate buffer" was one of the great mysteries of TSDB to me, and the insight that we need to cover one byte and that 4 samples is what it takes, even in the most compact encoding case, is IMHO a big revelation that should be documented.

codesome replied:
The latest update makes sense, thanks

codesome commented on tsdb/head.go:
```diff
@@ -1451,17 +1456,17 @@ type memSeries struct {
+	pendingCommit bool      // Whether there are samples waiting to be committed to this series.
 
-	app chunkenc.Appender // Current appender for the chunk.
+	app chunkenc.Appender // Current appender for the head chunk. Set when a new head chunk is cut. TODO when exactly is it nil or not nil? is it valid to be nil?
```
It is nil only if headChunk is nil. And that is possible if there was an appender that created a new series, but rolled back the commit (the first sample would create a headChunk, hence appender, but rollback skipped it while the Append() call would create a series).
```suggestion
	app chunkenc.Appender // Current appender for the head chunk. Set when a new head chunk is cut.
```

codesome commented on tsdb/head_append.go:
```diff
@@ -114,7 +114,7 @@ func (h *Head) Appender(_ context.Context) storage.Appender {
 
 func (h *Head) appender() *headAppender {
-	appendID, cleanupAppendIDsBelow := h.iso.newAppendID()
+	appendID, cleanupAppendIDsBelow := h.iso.newAppendID() // every appender gets an ID that is cleared again upon commit/rollback
```
```suggestion
	appendID, cleanupAppendIDsBelow := h.iso.newAppendID() // Every appender gets an ID that is cleared upon commit/rollback.
```

codesome commented on tsdb/head_append.go:
```diff
@@ -394,6 +395,7 @@ func exemplarsForEncoding(es []exemplarWithSeriesRef) []record.RefExemplar {
 	return ret
 }
 
+// Commit writes to the WAL and adds the data to the Head
```
```suggestion
// Commit writes to the WAL and adds the data to the Head.
```

codesome commented on tsdb/head_append.go:
```diff
@@ -343,6 +343,7 @@ func (a *headAppender) GetRef(lset labels.Labels) (uint64, labels.Labels) {
 	return s.ref, s.lset
 }
 
+// log writes all headAppender's data to the WAL
```
```suggestion
// log writes all headAppender's data to the WAL.
```

codesome commented on tsdb/head_append.go:
```diff
@@ -565,6 +567,7 @@ func (s *memSeries) mmapCurrentHeadChunk(chunkDiskMapper *chunks.ChunkDiskMapper
 	})
 }
 
+// Rollback removes the headAppender's data from the series (TODO is this correct) and writes the series to WAL.
```
> removes the headAppender's data from the series

This bit is not correct, it only discards the data from the appender (because we have not added to series yet at this point, it is only the isolation handling that you see below) while WAL logging the series records if any.
```suggestion
// Rollback removes the samples and exemplars from headAppender and writes the series to WAL if any.
```

codesome commented on tsdb/head_read.go:
```diff
@@ -175,6 +175,7 @@ func (h *headIndexReader) Series(ref uint64, lbls *labels.Labels, chks *[]chunks
 	return nil
 }
 
+// chunkID returns the ID corresponding to .mmappedChunks[pos]
```
```suggestion
// chunkID returns the ID corresponding to .mmappedChunks[pos] (head chunk if pos==len(mmappedChunks)).
```

codesome commented on tsdb/head_read.go:
```diff
@@ -351,7 +352,7 @@ func (c *safeChunk) Iterator(reuseIter chunkenc.Iterator) chunkenc.Iterator {
 }
 
-// iterator returns a chunk iterator.
+// iterator returns a chunk iterator for the requested chunkID // TODO I think?
```
That's correct
```suggestion
// iterator returns a chunk iterator for the requested chunkID.
```

codesome commented on tsdb/head_read.go:
```diff
@@ -455,6 +456,8 @@ func (s *memSeries) iterator(id int, isoState *isolationState, chunkDiskMapper *
 	}
 }
 
+// memSafe returns values from the wrapped stopIterator
```
```suggestion
// memSafeIterator returns values from the wrapped stopIterator
```

codesome commented on tsdb/head_read.go:
```diff
@@ -498,6 +501,8 @@ func (it *memSafeIterator) At() (int64, float64) {
 	return s.t, s.v
 }
 
+// stopIterator wraps an Iterator, but only returns the first
+// stopAfter-1 values. // TODO should this not be after stopAfter ?
```
It is indeed after stopAfter. `i` is initialised to -1 above. A newStopIterator() can make it more easily understandable in case you want to add, but out of scope of this PR .
```suggestion
// stopIterator wraps an Iterator, but only returns the first
// stopAfter values.
```

Dieterbe replied:
ah i see. it honors stopAfter if initialized with i=-1 (i assumed it would be initialized with the default zero value). I'll add a note for this.

bwplotka commented on tsdb/chunks/chunks.go:
```diff
@@ -68,7 +68,7 @@ type Meta struct {
 }
 
-// Iterator iterates over the chunk of a time series.
+// Iterator iterates over the chunks of a time series.
```
```suggestion
// Iterator iterates over the chunks of a single time series.
```

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -124,19 +124,20 @@ type ChunkDiskMapper struct {
 type mmappedChunkFile struct {
 	byteSlice ByteSlice
-	maxt      int64
+	maxt      int64 // maxt among all of this file's chunks.
```
```suggestion
	maxt      int64 // Max timestamp among all of this file's chunks.
```

bwplotka commented on tsdb/docs/usage.md:
```diff
@@ -0,0 +1,152 @@
+}
+
+func main() {
+	db, err := tsdb.Open("tsdb-test", nil, nil, tsdb.DefaultOptions(), nil)
```
This will age really badly. Why we cannot refer to tests? Or at least put it as `func Example()` in unit test suite?

bwplotka commented on tsdb/docs/usage.md:
````diff
@@ -0,0 +1,152 @@
+Example:
+
+```go
+package main
````
ditto

bwplotka reviewed (changes requested):
Amazing work. I would love to port those example to unit test of `func Example..`  to have it buildable and up to date. WDYT?

See https://go.dev/blog/examples

Dieterbe commented:
> I would love to port those example to unit test of func Example.. to have it buildable and up to date. WDYT?

@bwplotka I think it's a great idea, but as @roidelapluie said, godoc public docs for prometheus are currently out of date / for the wrong version.

Is there a link that I can/should use to point to the most current godoc for tsdb? Or can somebody fix the doc versioning issue? (it's out of scope for me).
Once this is resolved, i will be happy to move those code examples from the markdown into the code.

codesome commented on tsdb/docs/usage.md:
```diff
@@ -0,0 +1,152 @@
+
+## Instantiating a database
+
+Callers should use [`tsdb.Open`](https://pkg.go.dev/github.com/prometheus/prometheus/tsdb#Open) to open a TSDB.
+(the directory may be new or pre-existing).
```
```suggestion
Callers should use [`tsdb.Open`](https://pkg.go.dev/github.com/prometheus/prometheus/tsdb#Open) to open a TSDB
(the directory may be new or pre-existing).
```

codesome commented on tsdb/docs/usage.md:
```diff
@@ -0,0 +1,152 @@
+(the directory may be new or pre-existing).
+This returns a [`*tsdb.DB`](https://pkg.go.dev/github.com/prometheus/prometheus/tsdb#DB) which is the actual database.
+
+a `DB` has the following main components:
```
```suggestion
A `DB` has the following main components:
```

codesome commented on tsdb/docs/usage.md:
```diff
@@ -0,0 +1,152 @@
+
+a `DB` has the following main components:
+
+* Compactor: a [leveled compactor](https://github.com/prometheus/prometheus/blob/c0c22ed04200a8d24d1d5719f605c85710f0d008/tsdb/compact.go#L78). Note: it is currently the only compactor implementation. It runs automatically.
+* [`Head`](https://pkg.go.dev/github.com/prometheus/prometheus/tsdb#DB.Head)
```
There are also persistent blocks. You could point it to https://pkg.go.dev/github.com/prometheus/prometheus/tsdb#DB.Blocks.

Also, what was the reason to point leveled compactor it to the codebase and not https://pkg.go.dev/github.com/prometheus/prometheus/tsdb#LeveledCompactor?

Dieterbe replied:
> Also, what was the reason to point leveled compactor it to the codebase and not https://pkg.go.dev/github.com/prometheus/prometheus/tsdb#LeveledCompactor?

no good reason.  until we decide on how to address the godoc problem, i have them a bit intermixed.

codesome commented on tsdb/docs/usage.md:
```diff
@@ -0,0 +1,152 @@
+
+The `Head` is responsible for a lot.  Here are some of its main components:
+
+* [wal](https://pkg.go.dev/github.com/prometheus/prometheus/tsdb/wal#WAL) (Write Ahead Log)
+* [`stripeSeries`](https://github.com/prometheus/prometheus/blob/1270b87970baeb926fcce64552db5c744ffaf83f/tsdb/head.go#L1279):
+  this holds all the active series by linking to [`memSeries`](https://github.com/prometheus/prometheus/blob/1270b87970baeb926fcce64552db5c744ffaf83f/tsdb/head.go#L1449)
+  by an ID (aka "ref") and by labels hash.
+* postings list (reverse index): For any label-value pair, holds all the corresponding series refs. Used for queries.
+* tombstones
```
```suggestion
* [WAL](https://pkg.go.dev/github.com/prometheus/prometheus/tsdb/wal#WAL) (Write Ahead Log).
* [`stripeSeries`](https://github.com/prometheus/prometheus/blob/1270b87970baeb926fcce64552db5c744ffaf83f/tsdb/head.go#L1279):
  this holds all the active series by linking to [`memSeries`](https://github.com/prometheus/prometheus/blob/1270b87970baeb926fcce64552db5c744ffaf83f/tsdb/head.go#L1449)
  by an ID (aka "ref") and by labels hash.
* Postings list (reverse index): For any label-value pair, holds all the corresponding series refs. Used for queries.
* Tombstones.
```

codesome commented on tsdb/docs/usage.md:
```diff
@@ -0,0 +1,152 @@
+## Adding data
+
+Use [`db.Appender()`](https://pkg.go.dev/github.com/prometheus/prometheus/tsdb#DB.Appender) to obtain an "appender".
+The [golang docs](https://pkg.go.dev/github.com/prometheus/prometheus/storage#Appender) speak mostly for themselves.
```
uff, as pointed out previously, the golang docs have the wrong signature for the Appender. This could confuse developers more than it helps. Maybe we should fix the docs _before_ merging this PR (or make this and other godoc links point replaced with the codebase instead, pinned to the latest version and mention the version in this file).

codesome commented on tsdb/docs/usage.md:
```diff
@@ -0,0 +1,152 @@
+
+Remember:
+
+* Use Commit() to add the samples to the DB and update the WAL.
+* create a new appender each time you commit.
+* appenders are not concurrency safe, but scrapes run concurrently and as such, leverage multiple appenders concurrently.
```
```suggestion
* Use `Commit()` to add the samples to the DB and update the WAL.
* Create a new appender each time you commit.
* Appenders are not concurrency safe, but scrapes run concurrently and as such, leverage multiple appenders concurrently.
```

codesome commented on tsdb/docs/usage.md:
```diff
@@ -0,0 +1,152 @@
+
+1) `timestamp < minValidTime` where `minValidTime` is the highest of:
+  * the maxTime of the last block (i.e. the last truncation time of Head) - updated via [`Head.Truncate()`](https://pkg.go.dev/github.com/prometheus/prometheus/tsdb#Head.Truncate) and [`DB.CompactHead()`](https://github.com/prometheus/prometheus/blob/1270b87970baeb926fcce64552db5c744ffaf83f/tsdb/db.go#L968)
+  * `tsdb.min-block-duration/2` older than the max time in the Head chunk. Note that while technically `storage.tsdb.min-block-duration` is configurable, it's a hidden option and changing it is discouraged.  So We can assume this value to be 2h.
```
```suggestion
  * `tsdb.min-block-duration/2` older than the max time in the Head block. Note that while technically `storage.tsdb.min-block-duration` is configurable, it's a hidden option and changing it is discouraged.  So We can assume this value to be 2h.
```

codesome commented on tsdb/docs/usage.md:
```diff
@@ -0,0 +1,152 @@
+1) `timestamp < minValidTime` where `minValidTime` is the highest of:
+  * the maxTime of the last block (i.e. the last truncation time of Head) - updated via [`Head.Truncate()`](https://pkg.go.dev/github.com/prometheus/prometheus/tsdb#Head.Truncate) and [`DB.CompactHead()`](https://github.com/prometheus/prometheus/blob/1270b87970baeb926fcce64552db5c744ffaf83f/tsdb/db.go#L968)
+  * `tsdb.min-block-duration/2` older than the max time in the Head chunk. Note that while technically `storage.tsdb.min-block-duration` is configurable, it's a hidden option and changing it is discouraged.  So We can assume this value to be 2h.
+  Breaching this condition results in "out of bounds" errors.  
```
A newline above, else it is being rendered with the above point.
```suggestion

  Breaching this condition results in "out of bounds" errors.  
```

codesome commented on tsdb/docs/usage.md:
```diff
@@ -0,0 +1,152 @@
+  The first condition assures the block that will be generated doesn't overlap with the previous one (which simplifies querying)  
+  The second condition assures the sample won't go into the so called "compaction window", that is the section of the data that might be in process of being saved into a persistent block on disk.  (because that logic runs concurrently with ingestion without a lock)
+2) The labels don't validate. (if the set is empty or contains duplicate label names)
+3) If the sample, for the respective series (based on all the labels) is out of order or has a different value for the last (highest) timestamp seen. (results in storage.ErrOutOfOrderSample or storage.ErrDuplicateSampleForTimestamp respectively)
```
```suggestion
3) If the sample, for the respective series (based on all the labels) is out of order or has a different value for the last (highest) timestamp seen. (results in `storage.ErrOutOfOrderSample` and `storage.ErrDuplicateSampleForTimestamp` respectively)
```

codesome commented on tsdb/docs/usage.md:
```diff
@@ -0,0 +1,152 @@
+2) The labels don't validate. (if the set is empty or contains duplicate label names)
+3) If the sample, for the respective series (based on all the labels) is out of order or has a different value for the last (highest) timestamp seen. (results in storage.ErrOutOfOrderSample or storage.ErrDuplicateSampleForTimestamp respectively)
+
+Commit() may also refuse data that is out of order with respect to samples that were added via a different appender.
```
```suggestion
`Commit()` may also refuse data that is out of order with respect to samples that were added via a different appender.
```

codesome commented on tsdb/docs/usage.md:
```diff
@@ -0,0 +1,152 @@
+	ctx := context.Background()
+	app := db.Appender(ctx)
+
+	series := labels.FromStrings("foo", "bar")
+	ref, err := app.Append(0, series, time.Now().Unix(), 123)
+	time.Sleep(time.Second)
+	_, err = app.Append(ref, series, time.Now().Unix(), 124)
+	err = app.Commit()
+	noErr(err)
+	app = db.Appender(ctx) // need a new one. See app.Commit() docs
```
Added comments and minor edits to existing comments.

```suggestion
	series := labels.FromStrings("foo", "bar")
	// ref is 0 for the first append since we don't know the reference for the series.
	ref, err := app.Append(0, series, time.Now().Unix(), 123)
	time.Sleep(time.Second)
	// Re-using the ref from above since it's the same series, makes append faster.
	_, err = app.Append(ref, series, time.Now().Unix(), 124)
	err = app.Commit()
	noErr(err)
	app = db.Appender(ctx) // Need a new one. See app.Commit() docs.
```

codesome commented on tsdb/docs/usage.md:
```diff
@@ -0,0 +1,152 @@
+	db, err := tsdb.Open("tsdb-test", nil, nil, tsdb.DefaultOptions(), nil)
+	noErr(err)
+
+	querier, err := db.Querier(context.TODO(), math.MinInt64, math.MaxInt64)
```
Why context.TODO() here and context.Background() for the ingestion?

codesome reviewed (commented):
Generally happy with usage.md, most of my comments are minor nits.

One thing though: the godocs are outdated, and some interfaces have changed, so pointing to it here can confuse developers more than it helps. And also what @bwplotka mentioned about the code ageing badly.

But I don't thing either of us have time to fix and maintain the godoc updates (at least not right now? or is it simple to include in the CI pipeline?). So here is my suggestion to unblock this PR (while we try to fix godoc in parallel):
1. Pin the docs to the latest version by mentioning that this docs was last updated for v2.xx.x.
2. Replace all the current links (godoc and github) to point to the corresponding line of code for the latest comment in the respective release branch (example, for release `v2.31.x`, you will find a branch named `release-2.31`). PS: need not update links for new patch releases, they contain only bugfixes.

@bwplotka WDYT about this temporary solution?

codesome commented:
@Dieterbe let's go ahead with my suggestion in https://github.com/prometheus/prometheus/pull/9451#pullrequestreview-800390671 for now - mentioning v2.31.1 in the usage.md and replace all links (current github and godoc) to point to code lines in https://github.com/prometheus/prometheus/tree/411021ada9ab41095923b8d2df9365b632fd40c3

codesome commented:
Just found out with @Dieterbe that we can use the commit for godoc
Here is the one for v2.31.1: https://pkg.go.dev/github.com/prometheus/prometheus@v1.8.2-0.20211105201321-411021ada9ab

Dieterbe commented:
I have addressed all outstanding items. In particular pinned the docs/links and moved the example code to Example functions as part of the unit tests.

codesome commented on tsdb/example_test.go:
```diff
@@ -0,0 +1,104 @@
+// Copyright 2017 The Prometheus Authors
```
```suggestion
// Copyright 2021 The Prometheus Authors
```

codesome commented on README.md:
```diff
@@ -111,7 +111,7 @@ For more information on building, running, and developing on the new React-based
 ## More information
 
-  * The source code is periodically indexed: [Prometheus Core](https://pkg.go.dev/github.com/prometheus/prometheus).
+  * The source code is periodically indexed, but due to an issue with versioning, the "latest" docs shown on Godoc are outdated.                                                                                   Instead, you can use [the docs for v2.31.1](https://pkg.go.dev/github.com/prometheus/prometheus@v1.8.2-0.20211105201321-411021ada9ab).
```
```suggestion
  * The source code is periodically indexed, but due to an issue with versioning, the "latest" docs shown on Godoc are outdated. Instead, you can use [the docs for v2.31.1](https://pkg.go.dev/github.com/prometheus/prometheus@v1.8.2-0.20211105201321-411021ada9ab).
```

codesome reviewed (approved):
LGTM with nits! We should be linking example_test.go in the usage.md for it to be discoverable.

codesome commented on tsdb/example_test.go:
```diff
@@ -0,0 +1,104 @@
+	}
+}
+
+func Example() {
```
Since this is in _test.go, we can have it like a proper unit test, without panics, but with assertions

codesome commented:
(Fixed them myself, will merge on green)
